File: mailferry/engine/scheduler.py
"""
MailFerry - IMAP Migration & Sync
High-Performance Native IMAP Migration Engine

Copyright (C) 2026 Andy Saputra <andy@saputra.org>

https://saputra.org
https://github.com/ajsap/mailferry

Licensed under the GNU Affero General Public License v3.0 (AGPL-3.0).
This program is free software: you can redistribute it and/or modify it
under the terms of the GNU Affero General Public License as published by
the Free Software Foundation, either version 3 of the License, or (at
your option) any later version.

Contributions welcome: submit issues, feature requests and pull requests
at https://github.com/ajsap/mailferry

Batch scheduler: admission control, per-host connection budgets, dynamic
re-queueing (console `retry` / `reload`), graceful two-stage shutdown.
"""
from __future__ import annotations

import asyncio
import time
from typing import Dict, List

from ..config import MailboxSpec, RunConfig, parse_csv
from ..state.db import StateDB, lease_owner_id
from .mailbox import MailboxRunner


class HostSems:
    def __init__(self, per_host: int):
        self.per_host = max(1, per_host)
        self._sems: Dict[str, asyncio.Semaphore] = {}

    def sem(self, host: str) -> asyncio.Semaphore:
        h = host.lower()
        if h not in self._sems:
            self._sems[h] = asyncio.Semaphore(self.per_host)
        return self._sems[h]

    async def acquire_pair(self, a: str, b: str):
        for h in sorted((a.lower(), b.lower())):
            await self.sem(h).acquire()

    def release_pair(self, a: str, b: str):
        for h in (a.lower(), b.lower()):
            self.sem(h).release()


async def run_migration(cfg: RunConfig, specs: List[MailboxSpec], stats, session,
                        mailbox_logger_factory, stop_event: asyncio.Event,
                        hub=None) -> dict:
    loop = asyncio.get_event_loop()
    db = StateDB(cfg.db_path, ephemeral=cfg.ephemeral,
                 lease_fresh=getattr(cfg, "lock_timeout", StateDB.LEASE_FRESH))
    owner = lease_owner_id(cfg.run_id)
    await db.start_run(cfg.run_id, cfg.csv_file)
    await db.register_worker(owner, cfg.run_id)      # join the cluster
    if hub is not None:
        hub.worker_id = owner
        hub.add_history("Run started", "OK", "-",
                        f"{cfg.csv_file} · {len(specs)} mailbox(es) · run {cfg.run_id}")
        hub.add_history("Worker joined", "OK", "-",
                        f"{owner} — cluster on {cfg.db_path}")
        session.log(f"cluster: joined as worker {owner} "
                    f"(offline threshold {int(cfg.worker_timeout)}s)")

    pending: List[MailboxSpec] = []
    known_keys = set()
    for spec in specs:
        known_keys.add(spec.key)
        row = await db.upsert_mailbox(spec)
        mb = stats.mailbox(spec.index, spec.label, spec.src.host, spec.dst.host, spec.dst.user)
        if cfg.skip_completed and not cfg.force and row["status"] == "SUCCESS":
            mb.set(status="SKIPPED")
            stats.skipped_prior += 1
            continue
        if row["status"] == "RUNNING" and row["lease_owner"] and row["lease_owner"] != owner:
            age = time.time() - (row["lease_ts"] or 0)
            if age >= db.lease_fresh:
                session.log(f"note: {spec.src.user} was marked RUNNING by {row['lease_owner']} "
                            f"{int(age)}s ago (stale — that run crashed; resuming is correct)")
        pending.append(spec)

    if cfg.order == "size":
        sizes = {}
        for spec in pending:
            r = await db.upsert_mailbox(spec)
            sizes[spec.key] = r.get("bytes_total") or 0
        pending.sort(key=lambda s: -sizes.get(s.key, 0))

    workers = max(1, min(cfg.workers, len(pending))) if pending else 1
    stats.workers = workers
    worker_sem = asyncio.Semaphore(workers)
    hosts = HostSems(cfg.per_host_conns)
    extra_slots = asyncio.Semaphore(max(0, workers * (max(1, cfg.max_conns_per_mailbox) - 1)))
    tasks: List[asyncio.Task] = []

    async def run_one(spec: MailboxSpec):
        mb = stats.mailbox(spec.index, spec.label, spec.src.host, spec.dst.host, spec.dst.user)
        if hub is not None:
            hub.active_labels.add(spec.label)
        try:
            async with worker_sem:
                if stop_event.is_set():
                    mb.set(status="CANCELLED")
                    return
                await hosts.acquire_pair(spec.src.host, spec.dst.host)
                try:
                    runner = MailboxRunner(
                        spec, cfg, db, mb, session,
                        mailbox_logger_factory(spec), stop_event, owner,
                        extra_slots=extra_slots, host_sems=hosts, hub=hub)
                    await runner.run()
                except asyncio.CancelledError:
                    mb.set(status="CANCELLED", end_time=time.time())
                    raise
                except BaseException as e:      # a runner bug must not sink the batch
                    from ..errors import reason_of
                    mb.set(status="FAILED", error=f"internal error: {reason_of(e)}",
                           end_time=time.time())
                    session.log(f"[{spec.index:03d}] {spec.src.user}: INTERNAL ERROR {e!r}")
                    if cfg.debug:
                        raise
                finally:
                    hosts.release_pair(spec.src.host, spec.dst.host)
        finally:
            if hub is not None:
                hub.active_labels.discard(spec.label)

    def spawn(spec: MailboxSpec):
        mb = stats.mailbox(spec.index, spec.label, spec.src.host, spec.dst.host, spec.dst.user)
        mb.set(status="QUEUED", error="", end_time=0.0, retry_wait_until=0.0, attempt=1)
        tasks.append(loop.create_task(run_one(spec)))

    def reload_csv():
        added, known = 0, 0
        for spec in parse_csv(cfg.csv_file):
            if spec.key in known_keys:
                known += 1
                continue
            known_keys.add(spec.key)
            if hub is not None:
                hub.specs_by_label[spec.label] = spec
            spawn(spec)
            session.log(f"console: reload added {spec.src.user} -> {spec.dst.user}")
            added += 1
        return added, known

    if hub is not None:
        hub.db = db
        hub.spawn = spawn
        hub.reload_csv = reload_csv
        for spec in specs:
            hub.specs_by_label[spec.label] = spec

    for spec in pending:
        spawn(spec)

    supervisor_task = None
    if hub is not None and cfg.stale_timeout > 0:
        from .supervisor import stale_supervisor
        supervisor_task = loop.create_task(
            stale_supervisor(cfg, stats, hub, session, stop_event))
    monitor_task = None
    if hub is not None:
        from .cluster import cluster_monitor
        monitor_task = loop.create_task(
            cluster_monitor(cfg, stats, hub, session, db, owner, stop_event, spawn))

    try:
        while True:
            alive = [t for t in tasks if not t.done()]
            if alive:
                await asyncio.wait(alive, return_when=asyncio.FIRST_COMPLETED)
                continue
            if hub is not None and hub.remote_watch and not stop_event.is_set():
                # our local work is done but peers still hold mailboxes from
                # this project: stay alive — mirror their progress and stand
                # by for automatic failover (the monitor re-queues via spawn)
                try:
                    await asyncio.wait_for(stop_event.wait(), timeout=2.0)
                except asyncio.TimeoutError:
                    pass
                continue
            break
    finally:
        if supervisor_task is not None:
            supervisor_task.cancel()
        if monitor_task is not None:
            monitor_task.cancel()
        try:
            await db.deregister_worker(owner)        # graceful release
        except Exception:
            pass
        if hub is not None:
            hub.spawn = None                    # run is winding down
            hub.reload_csv = None
            if hub.shutdown_active:
                # scheduler stopped admitting; workers have now drained and
                # each runner already closed its connections with LOGOUT.
                hub.set_phase("scheduler", "done")
                hub.set_phase("workers", "done")
                hub.set_phase("conns", "done")
                hub.set_phase("state", "active")
        snap = stats.snapshot()
        counts = snap["counts"]
        result = (f"ok={counts.get('SUCCESS', 0)} partial={counts.get('PARTIAL', 0)} "
                  f"failed={counts.get('FAILED', 0)} stale={counts.get('STALE', 0)} "
                  f"cancelled={counts.get('CANCELLED', 0)}"
                  + (f" remote={counts.get('REMOTE', 0)}" if counts.get("REMOTE") else ""))
        if hub is not None:
            hub.add_history("Run finished", "OK", "-", result)
            try:
                hub.failed_registry = await db.failed_rows(
                    statuses=("FAILED", "RETRY_PENDING", "RETRYING"))
            except Exception:
                hub.failed_registry = []
        await db.end_run(cfg.run_id, result)
        db.close()
        if hub is not None and hub.shutdown_active:
            hub.set_phase("state", "done")       # State Database committed + closed

    for t in tasks:
        if not t.done():
            t.cancel()
    return stats.snapshot()

File: internal/engine/scheduler.go
// MailFerry — IMAP Migration & Sync
// A High-Performance Native IMAP Migration Engine
//
// Copyright (C) 2026 Andy Saputra
// Author: Andy Saputra <andy@saputra.org>
// SPDX-License-Identifier: AGPL-3.0-or-later
//
// This file is part of MailFerry (https://github.com/ajsap/mailferry).
// Licensed under the GNU Affero General Public License v3.0 or later;
// see the LICENSE file for details.

package engine

// Batch scheduler: bounded worker pool, per-host connection budgets,
// cluster worker registration, graceful cancellation.

import (
	"context"
	"fmt"
	"sort"
	"strings"
	"sync"
	"time"

	"github.com/ajsap/mailferry/v2/internal/config"
	"github.com/ajsap/mailferry/v2/internal/state"
)

type hostSems struct {
	mu      sync.Mutex
	perHost int
	sems    map[string]chan struct{}
}

func newHostSems(perHost int) *hostSems {
	if perHost < 1 {
		perHost = 1
	}
	return &hostSems{perHost: perHost, sems: map[string]chan struct{}{}}
}

func (h *hostSems) sem(host string) chan struct{} {
	h.mu.Lock()
	defer h.mu.Unlock()
	k := strings.ToLower(host)
	if h.sems[k] == nil {
		h.sems[k] = make(chan struct{}, h.perHost)
	}
	return h.sems[k]
}

func (h *hostSems) acquirePair(ctx context.Context, a, b string) bool {
	keys := []string{strings.ToLower(a), strings.ToLower(b)}
	if keys[1] < keys[0] {
		keys[0], keys[1] = keys[1], keys[0]
	}
	for i, k := range keys {
		select {
		case h.sem(k) <- struct{}{}:
		case <-ctx.Done():
			for j := 0; j < i; j++ {
				<-h.sem(keys[j])
			}
			return false
		}
	}
	return true
}

func (h *hostSems) releasePair(a, b string) {
	<-h.sem(a)
	<-h.sem(b)
}

type RunResult struct {
	Counts         map[string]int
	Outstanding    int64
	FailedRegistry []state.FailedRow
}

type LoggerFactory func(spec config.MailboxSpec) func(string, ...any)

// RunMigration executes the whole batch. ctx cancellation = graceful stop.
// bus may be nil (tests); a Bus wires history/logs/recovery to any UI.
func RunMigration(ctx context.Context, cfg *config.Run, specs []config.MailboxSpec,
	stats *Stats, session func(string), logFactory LoggerFactory) (RunResult, error) {
	return RunMigrationBus(ctx, cfg, specs, stats, NewBus(), session, logFactory)
}

func RunMigrationBus(ctx context.Context, cfg *config.Run, specs []config.MailboxSpec,
	stats *Stats, bus *Bus, session func(string), logFactory LoggerFactory) (RunResult, error) {
	db, err := state.Open(cfg.DBPath, cfg.Ephemeral, cfg.LockTimeout)
	if err != nil {
		return RunResult{}, err
	}
	defer db.Close()
	owner := state.LeaseOwnerID()
	bus.WorkerID = owner
	db.StartRun(cfg.RunID, cfg.CSVFile)
	db.RegisterWorker(owner, cfg.RunID)
	session(fmt.Sprintf("cluster: joined as worker %s (offline threshold %ds)",
		owner, int(cfg.WorkerTimeout)))
	bus.History("Run started", "OK", "-",
		fmt.Sprintf("%s · %d mailbox(es) · run %s", cfg.CSVFile, len(specs), cfg.RunID))
	bus.History("Worker joined", "OK", "-", owner+" — cluster on "+cfg.DBPath)
	defer db.DeregisterWorker(owner)

	hbStop := make(chan struct{})
	go func() {
		t := timeTicker(cfg.DBHeartbeat)
		defer t.Stop()
		for {
			select {
			case <-hbStop:
				return
			case <-t.C:
				db.WorkerHeartbeat(owner)
			}
		}
	}()
	defer close(hbStop)

	var pending []config.MailboxSpec
	for _, spec := range specs {
		mb := stats.Mailbox(spec.Index, spec.Label(), spec.Src.Host, spec.Dst.Host,
			spec.Dst.User)
		_, status, err := db.UpsertMailbox(spec.Key(), spec.Src.Host, spec.Src.User,
			spec.Dst.Host, spec.Dst.User)
		if err != nil {
			return RunResult{}, err
		}
		if cfg.SkipCompleted && !cfg.Force && status == "SUCCESS" {
			mb.Set(func(m *MBValues) { m.Status = "SKIPPED" })
			continue
		}
		pending = append(pending, spec)
	}

	if cfg.Order == "size" {
		// Largest known mailbox first (unknown sizes keep CSV order at the end).
		sort.SliceStable(pending, func(i, j int) bool {
			return db.SizeOfKey(pending[i].Key()) > db.SizeOfKey(pending[j].Key())
		})
	}

	workers := cfg.Workers
	if workers > len(pending) {
		workers = len(pending)
	}
	if workers < 1 {
		workers = 1
	}
	stats.mu.Lock()
	stats.Workers = workers
	stats.mu.Unlock()

	hosts := newHostSems(cfg.PerHostConns)
	slots := make(chan struct{}, workers)
	reg := NewRegistry()
	watch := newRemoteWatch()
	var wg sync.WaitGroup

	var runOne func(spec config.MailboxSpec)
	runOne = func(spec config.MailboxSpec) {
		defer wg.Done()
		select {
		case slots <- struct{}{}:
			defer func() { <-slots }()
		case <-ctx.Done():
			stats.Mailbox(spec.Index, spec.Label(), spec.Src.Host, spec.Dst.Host,
				spec.Dst.User).Set(func(m *MBValues) { m.Status = "CANCELLED" })
			return
		}
		if ctx.Err() != nil {
			stats.Mailbox(spec.Index, spec.Label(), spec.Src.Host, spec.Dst.Host,
				spec.Dst.User).Set(func(m *MBValues) { m.Status = "CANCELLED" })
			return
		}
		if !hosts.acquirePair(ctx, spec.Src.Host, spec.Dst.Host) {
			stats.Mailbox(spec.Index, spec.Label(), spec.Src.Host, spec.Dst.Host,
				spec.Dst.User).Set(func(m *MBValues) { m.Status = "CANCELLED" })
			return
		}
		defer hosts.releasePair(spec.Src.Host, spec.Dst.Host)
		mb := stats.Mailbox(spec.Index, spec.Label(), spec.Src.Host, spec.Dst.Host,
			spec.Dst.User)
		runner := &MailboxRunner{Spec: spec, Cfg: cfg, DB: db, MB: mb, Bus: bus,
			Session: session, Logf: logFactory(spec), Ctx: ctx, Owner: owner}
		reg.Add(spec.Label(), runner)
		st := runner.Run()
		reg.Remove(spec.Label())
		if st == "REMOTE" && ctx.Err() == nil {
			watch.Add(spec, runner.MID, "")
			if o, _ := db.ReadLease(runner.MID); o != "" {
				watch.Add(spec, runner.MID, o)
			}
		}
	}
	respawn := func(spec config.MailboxSpec) {
		wg.Add(1)
		go runOne(spec)
	}
	for _, spec := range pending {
		spec := spec
		wg.Add(1)
		go runOne(spec)
	}

	// Control hooks for the presentation layer (shutdown escalation, manual
	// retry, CSV reload). The engine stays authoritative: hooks re-enter the
	// same admission path as everything else.
	bus.SetAborter(func() int {
		n := reg.AbortAll()
		if n > 0 {
			session(fmt.Sprintf("shutdown: closed %d IMAP connection(s) to stop promptly", n))
		}
		return n
	})
	specByLabel := map[string]config.MailboxSpec{}
	for _, sp := range specs {
		specByLabel[sp.Label()] = sp
	}
	bus.SetRequeue(func(label string, all bool) int {
		n := 0
		for _, m := range stats.Snapshot().Mailboxes {
			if m.Status != "FAILED" && m.Status != "PARTIAL" && m.Status != "STALE" {
				continue
			}
			if !all && m.Label != label {
				continue
			}
			sp, okSp := specByLabel[m.Label]
			if !okSp || reg.Get(m.Label) != nil {
				continue
			}
			bus.ClearStale(m.Label)
			bus.Log("INFO", m.Label, "re-queued for retry")
			bus.History("Mailbox re-queued", "OK", m.Label, "manual retry from the console")
			respawn(sp)
			n++
		}
		return n
	})
	bus.SetReloader(func() (int, int) {
		newSpecs, err := config.ParseCSV(cfg.CSVFile)
		if err != nil {
			bus.Log("WARN", "-", "CSV reload failed: "+err.Error())
			return 0, len(specByLabel)
		}
		added := 0
		for _, sp := range newSpecs {
			if _, known := specByLabel[sp.Label()]; known {
				continue
			}
			sp.Index = len(specByLabel) + added + 1
			specByLabel[sp.Label()] = sp
			db.UpsertMailbox(sp.Key(), sp.Src.Host, sp.Src.User, sp.Dst.Host, sp.Dst.User)
			bus.History("Mailbox added", "OK", sp.Label(), "picked up from the CSV")
			respawn(sp)
			added++
		}
		if added > 0 {
			session(fmt.Sprintf("csv reload: %d new mailbox(es) admitted", added))
		}
		return added, len(newSpecs)
	})

	supCtx, supCancel := context.WithCancel(context.Background())
	defer supCancel()
	go StaleSupervisor(supCtx, cfg, stats, bus, reg, session)
	go clusterMonitor(supCtx, cfg, stats, bus, db, owner, session, watch, respawn)

	// wait for local work; while peers still hold project mailboxes, stay
	// alive — mirror their progress and stand by as a hot failover
	done := make(chan struct{})
	go func() { wg.Wait(); close(done) }()
	for {
		select {
		case <-done:
		case <-time.After(2 * time.Second):
			if watch.Len() > 0 && ctx.Err() == nil {
				continue
			}
			select {
			case <-done:
			default:
				continue
			}
		}
		if watch.Len() > 0 && ctx.Err() == nil {
			select {
			case <-ctx.Done():
			case <-time.After(2 * time.Second):
			}
			continue
		}
		break
	}
	wg.Wait()

	snap := stats.Snapshot()
	agg := snap.Agg()
	res := RunResult{Counts: agg.Counts, Outstanding: db.OutstandingFailed(),
		FailedRegistry: db.FailedRows(0, true)}
	result := fmt.Sprintf("ok=%d warnings=%d partial=%d failed=%d stale=%d cancelled=%d",
		agg.Counts["SUCCESS"], agg.Counts["WARNINGS"], agg.Counts["PARTIAL"],
		agg.Counts["FAILED"], agg.Counts["STALE"], agg.Counts["CANCELLED"])
	bus.History("Run finished", "OK", "-", result)
	db.EndRun(cfg.RunID, result)
	return res, nil
}

func timeTicker(secs float64) *tickerWrap {
	if secs < 5 {
		secs = 5
	}
	return newTicker(secs)
}
